src/dsa/rpo_stark/stark/mod.rs

```diff
@@ -0,0 +1,80 @@
+use core::marker::PhantomData;
+
+use prover::RpoSignatureProver;
+use rand::{distributions::Standard, prelude::Distribution, thread_rng, RngCore, SeedableRng};
+use rand_chacha::ChaCha20Rng;
+use winter_crypto::{ElementHasher, Hasher, SaltedMerkleTree};
+use winter_math::{fields::f64::BaseElement, FieldElement};
+use winter_prover::{Proof, ProofOptions};
+use winter_verifier::{verify, AcceptableOptions, VerifierError};
+use winterfell::Prover;
+
+use crate::{
+    hash::rpo::{Rpo256, DIGEST_RANGE, DIGEST_SIZE, NUM_ROUNDS, STATE_WIDTH},
+    rand::RpoRandomCoin,
+};
+
+mod air;
+pub use air::{PublicInputs, RescueAir};
+mod prover;
+
+/// Represents an abstract STARK-based signature scheme with knowledge of RPO pre-image as
+/// the hard relation.
+pub struct RpoSignatureScheme<H: ElementHasher> {
+    options: ProofOptions,
+    _h: PhantomData<H>,
+}
+
+impl<H: ElementHasher<BaseField = BaseElement> + Sync> RpoSignatureScheme<H>
+where
+    Standard: Distribution<<H as Hasher>::Digest>,
+{
+    pub fn new(options: ProofOptions) -> Self {
+        RpoSignatureScheme { options, _h: PhantomData }
+    }
+
+    pub fn sign(&self, sk: [BaseElement; DIGEST_SIZE], msg: [BaseElement; DIGEST_SIZE]) -> Proof {
+        // create a prover
+        let prover = RpoSignatureProver::<H>::new(self.options.clone());
+
+        // generate execution trace
+        let trace = prover.build_trace(sk, msg);
+
+        // generate the initial seed for the PRNG used for zero-knowledge
+        let mut seed = <ChaCha20Rng as SeedableRng>::Seed::default();
+        let mut rng = thread_rng();
+        rng.fill_bytes(&mut seed);
```

Comment:
We shouldn't be able to do this in WASM (or any no-std context), right? I wonder what else can we do here (having a fully deterministic signature would have been nice - but we don't have this yet, right?)

Reply:
I think I have now made everything depend on one seed, except for the salted Merkle tree which we should discuss in the relevant issue.
Once this is fixed we should be able to de-randomize the signature e.g., make the initial seed a deterministic function of the inputs and some other fixed constants